Add support for building in libvirt Vagrant instance

bkus opened the pull request:
Doing this to ultimately support libvirt RouterOS instances.

bkus committed:
Add support for building in libvirt Vagrant instance

Doing this to ultimately support libvirt RouterOS instances.

cheretbe commented:
Hi @bkus 
Are you going to implement full libvirt support eventually?
I was planning to, but right now I don't use RouterOS very much and it is quite low in the list of my priorities.

bkus committed:
Move plugin builds on libvirt to Fedora 35

This is the last version of Fedora that provided a ruby 2.7 package.

Unable to make snap builds work due to either dynamic linker errors, or apparmor
violations.  Unable to disable apparmor enforcement since the rules come in via
snap, but #include <> stuff from global scope, and the include paths aren't
compatible.  Unable to unload apparmor module since Ubuntu appears to have it
built directly into kernel.  Stopping the apparmor service only stops ability
to update rules, but doesn't disable the already-configured ones from breaking
builds.

A "sudo BUNDLE_SILENCE_ROOT_WARNING=true bundle install" is required in the VM
since some gems build native extensions, and those aren't installed or detected
properly by the subsequenct "bundle exec rake build".

bkus commented:
@cheretbe Yeah I'm trying to make a libvirt compatible RouterOS box from this, so I can test HamWAN Ansible automation against instances of it.  If you're willing to keep reviewing / merging, I can keep producing changes that are compatible with your original VirtualBox implementation of this.  If you're too busy or whatever, it'd be a lot easier on me to just drop the VBox stuff and all the associated conditionals, and just reduce the HamWAN fork of this to libvirt only.

Let me know what you wanna do!

And thanks for creating this in the first place.  :)

bkus committed:
Automatic PEP8 style update with Black and isort

https://github.com/psf/black
https://github.com/pycqa/isort

bkus committed:
Removing PyInquirer dependency, and updating tasks with sudo bundle

There's a fix for the PyInquirer breakage:
  https://github.com/CITGuru/PyInquirer/pull/194
But PyInquirer doesn't have a maintainer anymore:
  https://github.com/CITGuru/PyInquirer/issues/159

The sudo bundle update is because some gems have native extensions, and those
don't get installed or recognized properly by the subsequent rake build.

bkus committed:
Simplify the build instructions with less redirection

bkus committed:
Minor formatting and doc updates

bkus committed:
Enable building of Packer boxes via QEMU provider

QEMU is auto-detected, and used if present.  Virtualbox is fallback.

cheretbe commented:
@bkus I definitely would prefer making changes in this repo and in the end to have fully functional boxes that support both virtualbox and libvirt.
Right now I cannot test anything in libvirt environment (this may change within a month's time). Yes, I'm willing to review/merge changes as long as they will not break existing virtualbox functionality.

Thank you for taking interest is this project. It would be nice to make it a bit more usable.

Do you have experience with Ruby? I'm asking because I'd like to separate the `vagrant-routeros` plugin and publish it on [rubygems.org](https://rubygems.org/). But first it needs to have proper network configuration support. This would require deeper knowledge of inner Vagrant plugins mechanics than I have.

bkus commented:
@cheretbe Good to hear you're willing to take the changes into this repo.  I think everything to date hasn't broken the vbox functionality, although I haven't setup any test env for that.  Let me know where your SSH key is if you need some server resources, I have lots of hypervisors available.

I only started touching Ruby 2-3 weeks ago, when I had to write some Vagrantfiles.  I'll need to learn more of it though, to complete the RouterOS hooks.

Now that I have the plugin building in libvirt-qemu-kvm and the boxes building in qemu-kvm, today's battle will be with getting the boxes to run in a libvirt-vagrant and getting the tests to pass.  This pull request is getting quite large though.  Not sure how you wanna handle the merges.  Let me know if I should stack a new branch & PR on top of this one, or something else.  Otherwise, I'll just keep adding commits to this branch.

bkus committed:
Enable concurrent box builds

I've noticed the terminal goes into "invisible typing" mode with
concurrent packer outputs.  Not sure how to solve this.  A simple
"reset" command does clear it up though.

bkus committed:
Revert "Enable concurrent box builds"

This reverts commit 969e00590aabcc40d153e3e42aee21be1de2742d.

There are still problems with packer launching multiple instances, and I
can't find them easily.  Given this and the terminal corruption, it's probably
best to abandon concurrency in the build task and instead come up with
something that uses packer's native parallel build support:

$ packer build [-only=...|-except=...] -parallel-builds=0

cheretbe commented:
@bkus I think a new branch is a good option. At least that's what I usually do on gitlab (separate feature branch + MR). I don't do much collaboration on github, since I use it for my personal things, so if you prefer different workflow just let me know.
I've created `libvirt-build-support` branch just in case.
Yes, I'd like to do some testing on a KVM hypervisor if possible. Will not have time for this the following couple of weeks though, but sent you my public SSH key just to set up things in advance.

bkus committed:
Implement Packer-native concurrent builds

HCL2 appears to lack the expressiveness to generate top level blocks on
its own, such as "source" and "build".  While it does have some loop
support, it appears limited to generating collections.  Had to create a
jinja2 template for the HCL2 template in order to get all the builds
expressed simultaneously and allow Packer to utilize its parallelism.

There is a new "inv template" task that can be run manually, and it's
also hooked as a pre-task to all routeros build tasks.

Also decided to change the initialization of the context to prepopulate
it with the latest RouterOS versions from Mikrotik.  This data is used
throughout the build program, and it was being fetched in multiple
redundant places.  Now it's centralized and always available.

Had to change the box filenames to include a provider in the middle.
The new syntax is: <branch>_<provider>_<version>.box.  Since we're
supporting multiple build providers now, the filename distinction needs
to be made.

Dropped the generation of the associated *.md files along with the
boxes.  Not sure what value it provided, since the boxes already carry
their own unique version information.

Finally, removed the ask_for_confirmation when invoking a build.  If
someone didn't want to build, they wouldn't have invoked the command.

Future note: if the build host is equipped with both libvirt and
virtualbox, it may be possible to run builds in both systems
concurrently.  This would produce all possible boxes in the shortest
amount of time.  The current logic of selecting libvirt in favor of
virtualbox may need to be changed to "select all available providers".

bkus commented:
@cheretbe Where's your public key?  I don't see anything.

I think I changed my mind about the branching.  The main problem is I'm writing a ton of stacked changes, and as long as they don't get merged they're just gonna keep stacking up, no matter if I add extra branches to increase complexity or not.  I'll keep adding changes to this branch, and you can cherry-pick the commits if you want smaller units of change.  I'm trying to keep the commits to a reasonable size, although this last one is a doozy.  :)  On the plus side, 3 boxes are building in about 2.5 minutes.

cheretbe commented:
@bkus I emailed the key to the address in your Github profile. I feel a little bit uneasy sharing it publicly, even though it should be safe.

As for branching. I'm ok with merging changes of any size. Let's not over complicate things. After all, this project is not that big :smile:

bkus commented:
@cheretbe Big problem.  As I'm starting the ROS VMs under libvirt/qemu/kvm now, all I'm getting is "Demo license expired!  Please reinstall the router" followed by immediate shutdown.  Looking at Mikrotik's licensing (https://wiki.mikrotik.com/wiki/Manual:License#License_Levels) it appears x86 Trial licenses only last for 24 hours after install.  When I rebuild the boxes and start them < 24hrs later, I get the same error, so there's some bug there.  But even if that were solved somehow, wouldn't the published boxes all be useless after 24 hours anyways?

I see the CHR flavor of RouterOS does offer a perpetual free license (https://wiki.mikrotik.com/wiki/Manual:CHR#free), albeit limited to 1Mbit interface output speeds.  For Vagrant and testing, this seems like a better choice.  What do you think of migrating the project to distribute CHR images instead?  It would eliminate the ISO install process, so it's a pretty big change.

I'm not sure how much CHR differs from x86 ROS in terms of commands / API.

bkus commented:
@cheretbe Maybe https://github.com/vagrant-libvirt/vagrant-libvirt/pull/1715 will help.  If we can start the boxes with a fixed real time clock value (absolute time), they might be useful for 24 hours each time.

cheretbe commented:
@bkus Yeah, ROS licensing policy is very complicated. My primary use case for these boxes is testlabs. They are temporary by definition and designed so that it's easy to recreate them at any time. I do the real configuration on hardware boxes, that have preinstalled licences. That's why I didn't look into CHR, even though I've heard of it.
Will check it, may be we can offer it as an additional option

bkus committed:
Migrate the plugin build process out of Vagrant

This is many times faster than with a Vagrant build VM.  The dependency on
Ruby 2.7 is also eliminated with this change.  Removed manual build
instructions and replaced with a reference to more granular build tasks.
The new build system uses a localized (not global) bundle installation.
This will keep the global build environment clean.  Global build requirements
have been updated in the README.md.  Debian/Ubuntu build environment
instructions should be possible, but are not provided at this time.

cheretbe commented:
@bkus Looks like Mikrotik provides ready-to-use with KVM disk images for CHR:
* https://wiki.mikrotik.com/wiki/Manual:CHR_ProxMox_installation

cheretbe reviewed (commented):
I'm not sure if removing manual mode with `-on-error` action is a good idea. I added it for a reason: when there is a problem with the build, manual steps and ability to log into the VM after failure helps to debug the issue.

bkus commented:
Yeah, you're right.  `-on-error` should be configurable to hold the VM(s) alive.  I removed `--batch` cuz with the elimination of the plugin build VM, `-on-error` was the only place left it was used and I was always using --batch mode.  I'll re-introduce some flag to configure a manual `-on-error` mode.

bkus committed:
Introduce --debug flag to build tasks

This allows failing Packer builds to hold their VMs alive, so that they
may be accessed for debugging.
